File: ruleMatch/src/main/scala/com/bigdata/rulematch/scala/function/EventJSONToBeanFlatMapFunction.scala
package com.bigdata.rulematch.scala.function

import com.alibaba.fastjson.JSON
import com.bigdata.rulematch.scala.bean.EventLogBean
import org.apache.flink.api.common.functions.RichFlatMapFunction
import org.apache.flink.util.Collector
import org.slf4j.{Logger, LoggerFactory}

/**
 *
 * 把event事件的json字符串格式转换为一个EventLogBean对象
 *
 * @author Administrator
 * @version 1.0
 * @date 2021-12-18 19:23
 */
class EventJSONToBeanFlatMapFunction extends RichFlatMapFunction[String, EventLogBean] {
  private val logger: Logger = LoggerFactory.getLogger(this.getClass.getName.stripSuffix("$"))

  override def flatMap(value: String, out: Collector[EventLogBean]): Unit = {

    try {
      //将JSON 数据变成 EventLogBean
      val eventLogBean = JSON.parseObject(value, classOf[EventLogBean])

      out.collect(eventLogBean)
    } catch {
      case e: Throwable => {
        logger.error(s"JSON解析异常: ${value}", e)
      }
    }


  }
}

File: ruleMatch/src/main/java/com/bigdata/rulematch/java/function/EventJSONToBeanFlatMapFunction.java
package com.bigdata.rulematch.java.function;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.bigdata.rulematch.java.bean.EventLogBean;
import org.apache.flink.api.common.functions.RichFlatMapFunction;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author HanDongfang
 * @create 2021-12-19  22:02
 */
public class EventJSONToBeanFlatMapFunction extends RichFlatMapFunction<String, EventLogBean> {
    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());

    @Override
    public void flatMap(String s, Collector<EventLogBean> collector) throws Exception {
        try {

            JSONObject jsonObject = JSON.parseObject(s);
            /*String userId = jsonObject.getString("userId");
            String eventId = jsonObject.getString("eventId");
            Long timeStamp = jsonObject.getLong("timeStamp");
            String properties = jsonObject.getString("properties");*/
            //将JSON 数据变成 EventLogBean
            EventLogBean eventLogBean = JSON.toJavaObject(jsonObject, EventLogBean.class);

            collector.collect(eventLogBean);

        } catch (Exception e) {
            logger.warn("数据格式出错: " + s);
        }
    }
}
